### PR TITLE
Add the `logfire.instrument_mysql()`

### LINKED ISSUE #310: Add the `logfire.instrument_mysql()`
### Description

OpenTelemetry MySQL Instrumentation[](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/mysql/mysql.html#module-opentelemetry.instrumentation.mysql)

### PR DESCRIPTION
**Pull request by aditkumar72**
Fixes #310

### CODE REVIEW

**Comment by aditkumar72** [comment]
@Kludex Could you please review the PR ?

**Review by Kludex — COMMENTED** [review]
On this integration, you can also [instrument a single connection](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/mysql/mysql.html#opentelemetry.instrumentation.mysql.MySQLInstrumentor.instrument_connection).

@alexmojaki Can you confirm we want this to be analogous to psycopg?

**Comment by alexmojaki** [comment]
We do want that, but I don't know if that needs to block this PR.

**Comment by aditkumar72** [comment]
@Kludex @alexmojaki I will update this PR as this is not an urgent required changes

File: logfire/__init__.py [hunk 1]
```diff
@@ -112,6 +113,7 @@ def loguru_handler() -> dict[str, Any]:
     'instrument_sqlalchemy',
     'instrument_redis',
     'instrument_pymongo',
+    'instrument_mysql',
```
Comment (alexmojaki) [inline logfire/__init__.py:116]: Also needs to be added to `__all__` in `logfire_api/__init__.pyi`

File: logfire/_internal/main.py [hunk 2]
```diff
@@ -739,7 +756,7 @@ def with_settings(
             tags=self._tags + tuple(tags),
             sample_rate=self._sample_rate,
             console_log=self._console_log if console_log is None else console_log,
-            otel_scope=self._otel_scope if custom_scope_suffix is None else f'logfire.{custom_scope_suffix}',
+            otel_scope=(self._otel_scope if custom_scope_suffix is None else f'logfire.{custom_scope_suffix}'),
```
Comment (alexmojaki) [inline logfire/_internal/main.py]: I'm seeing a lot of formatting changes like this that are unrelated to the PR. Please check that you're using `ruff==0.5.4` for formatting and revert these changes.

File: logfire/_internal/main.py [hunk 3]
```diff
@@ -1223,6 +1263,30 @@ def instrument_redis(self, **kwargs: Unpack[RedisInstrumentKwargs]) -> None:
         self._warn_if_not_initialized_for_instrumentation()
         return instrument_redis(**kwargs)
 
+    def instrument_mysql(
+        self,
+        conn: PooledMySQLConnection | MySQLConnectionAbstract | None = None,
+        **kwargs: Unpack[MySQLInstrumentKwargs],
+    ) -> PooledMySQLConnection | MySQLConnectionAbstract | None:
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        conn: MySQLConnection = None,
        **kwargs: Unpack[MySQLInstrumentKwargs],
    ) -> MySQLConnection:
```

where `MySQLConnection = TypeVar('MySQLConnection', PooledMySQLConnection, MySQLConnectionAbstract, None)`

File: tests/otel_integrations/test_mysql.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import mysql.connector
+import pytest
+from inline_snapshot import snapshot
+
+import logfire
+from logfire.testing import TestExporter
+
+
+@pytest.fixture(autouse=True)
+def mock_mysql_connection():
```
Comment (alexmojaki) [inline tests/otel_integrations/test_mysql.py]: Please use https://testcontainers-python.readthedocs.io/en/latest/modules/mysql/README.html

**Review by alexmojaki — CHANGES_REQUESTED** [review]

**Comment by aditkumar72** [comment]
Hi @alexmojaki , resolved the comments. Could you please review ?

File: pyproject.toml [hunk 5]
```diff
@@ -136,6 +138,9 @@ dev-dependencies = [
     "mypy>=1.10.0",
     "celery>=5.4.0",
     "testcontainers",
+    "mysql-connector-python~=8.0",
+    "pymysql",
+    "cryptography",
```
Comment (alexmojaki) [inline pyproject.toml]: I see `mysql-connector-python` being used, but what about these 2?

File: tests/otel_integrations/test_mysql.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                    'net.peer.port': IsInt(),
+                },
+            },
+        ]
+    )
+
+
+def test_instrument_mysql_connection(exporter: TestExporter, mysql_container: MySqlContainer):
+    conn = get_mysql_connection(mysql_container)
+    conn = logfire.instrument_mysql(conn)  # type: ignore
```
Comment (alexmojaki) [inline tests/otel_integrations/test_mysql.py]: Why the `type: ignore`? We don't want users to have to write that.

Please extend this test to show that only this specific connection is being instrumented. You will probably have to uninstrument the module at the end of `test_mysql_instrumentation` to ensure that another uninstrumented connection doesn't create spans.

**Review by alexmojaki — CHANGES_REQUESTED** [review]

File: logfire/_internal/main.py [hunk 7]
```diff
@@ -1223,6 +1224,30 @@ def instrument_redis(self, **kwargs: Unpack[RedisInstrumentKwargs]) -> None:
...
+        [OpenTelemetry MySQL Instrumentation](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/mysql/mysql.html)
+        library.
+
+        Args:
+            conn: The `mysql` connection to instrument, or `None` to instrument all connections.
+            **kwargs: Additional keyword arguments to pass to the OpenTelemetry `instrument` methods.
+
+        Returns:
+            If a connection is provided, returns the instrumented connection. If no connection is provided, returns None.
+
```
Comment (Kludex) [inline logfire/_internal/main.py:1244]:
```suggestion
```

File: logfire/_internal/integrations/mysql.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    This function uses the OpenTelemetry MySQL Instrumentation library to instrument either the entire `mysql` module or a specific MySQL connection.
+
+    Args:
+        conn: The MySQL connection to instrument. If None, the entire `mysql` module is instrumented.
+        **kwargs: Additional keyword arguments to pass to the OpenTelemetry `instrument` methods.
+
+    Returns:
+        If a connection is provided, returns the instrumented connection. If no connection is provided, returns None.
+
```
Comment (Kludex) [inline logfire/_internal/integrations/mysql.py:32]:
```suggestion

```

File: logfire-api/logfire_api/_internal/main.pyi [hunk 9]
```diff
@@ -618,6 +619,22 @@ class Logfire:
...
+        [OpenTelemetry MySQL Instrumentation](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/mysql/mysql.html)
+        library.
+
+        Args:
+            conn: The `mysql` connection to instrument, or `None` to instrument all connections.
+            **kwargs: Additional keyword arguments to pass to the OpenTelemetry `instrument` methods.
+
+        Returns:
+            If a connection is provided, returns the instrumented connection. If no connection is provided, returns None.
+
```
Comment (Kludex) [inline logfire-api/logfire_api/_internal/main.pyi:636]:
```suggestion
```

File: logfire-api/logfire_api/_internal/integrations/mysql.pyi [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    This function uses the OpenTelemetry MySQL Instrumentation library to instrument either the entire `mysql` module or a specific MySQL connection.
+
+    Args:
+        conn: The MySQL connection to instrument. If None, the entire `mysql` module is instrumented.
+        **kwargs: Additional keyword arguments to pass to the OpenTelemetry `instrument` methods.
+
+    Returns:
+        If a connection is provided, returns the instrumented connection. If no connection is provided, returns None.
+
```
Comment (Kludex) [inline logfire-api/logfire_api/_internal/integrations/mysql.pyi:32]:
```suggestion

```

File: tests/otel_integrations/test_mysql.py (hunk 6 above)
Comment (aditkumar72) [reply tests/otel_integrations/test_mysql.py]: `type: ignore` was added because `get_mysql_connection` return type was not mentioned. If we add `MySQLConnection` as return type then we are getting type error:

```
Expression of type "PooledMySQLConnection | MySQLConnectionAbstract" is incompatible with return type "MySQLConnection@get_mysql_connection"   Type "PooledMySQLConnection | MySQLConnectionAbstract" is incompatible with type "MySQLConnection@get_mysql_connection"
```

 Fixed it with `Any`

File: pyproject.toml (hunk 5 above)
Comment (aditkumar72) [reply pyproject.toml]: These dependencies are required for using MySqlContainer in Python 3.8

**Comment by aditkumar72** [comment]
@alexmojaki Could you please review the changes ?

**Review by Kludex — CHANGES_REQUESTED** [review]

File: logfire/_internal/main.py (hunk 7 above)
Comment (alexmojaki) [reply logfire/_internal/main.py:1244]: I think this is fine. We have something similar in instrument_psycopg. All of this is useful, especially the fact that it returns an instrumented connection.

File: tests/otel_integrations/test_mysql.py (hunk 6 above)
Comment (alexmojaki) [reply tests/otel_integrations/test_mysql.py]: This test still needs to show that an uninstrumented connection doesn't create spans.

**Comment by aditkumar72** [comment]
@Kludex please check this comment https://github.com/pydantic/logfire/pull/341#discussion_r1700471697
Do we need to remove those docstrings ?

**Review by alexmojaki — APPROVED** [review]

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
Thanks @aditkumar72 🙏
